python/global_search_completer/global_search_completer.py

```diff
@@ -0,0 +1,362 @@
+
+    def __on_worker_failure(self, uid, msg):
+        """
+        Asynchronous callback - the worker thread errored.
+
+        :param uid: Unique id for request that failed
+        :param msg: Error message
+        """
+        uid = shotgun_model.sanitize_qt(uid) # qstring on pyqt, str on pyside
+        msg = shotgun_model.sanitize_qt(msg)
+        if self._processing_id == uid:
+            self._bundle.log_warning("Could not retrieve search results: %s" % msg)
+
+    def __on_worker_signal(self, uid, request_type, data):
+        """
+        Signaled whenever the worker completes something.
+        This method will dispatch the work to different methods
+        depending on what async task has completed.
+
+        :param uid: Unique id for request
+        :param request_type: String identifying the request class
+        :param data: the data that was returned
+        """
+        uid = shotgun_model.sanitize_qt(uid) # qstring on pyqt, str on pyside
+        data = shotgun_model.sanitize_qt(data)
+
+        if uid in self._thumb_map:
+            thumbnail = data["image"]
+            if thumbnail:
+                item = self._thumb_map[uid]["item"]
+                thumb = QtGui.QPixmap.fromImage(thumbnail)
+                item.setIcon(thumb)
+
+
+        if self._processing_id == uid:
+            # all done!
+            self._clear_model(add_loading_item=False)
+
+            matches = data["return_value"]["matches"]
+
+            if len(matches) == 0:
+                item = QtGui.QStandardItem("No matches found!")
+                item.setData(self.MODE_NOT_FOUND, self.MODE_ROLE)
+                self.model().appendRow(item)
+
+            # insert new data into model
+            for d in matches:
+                item = QtGui.QStandardItem(d["name"])
+                item.setData(self.MODE_RESULT, self.MODE_ROLE)
+
+                item.setData(shotgun_model.sanitize_for_qt_model(d),
+                             self.SG_DATA_ROLE)
+
+                item.setIcon(self._default_icon)
+
+                if d.get("image", None) and self.__sg_data_retriever:
+                    uid = self.__sg_data_retriever.request_thumbnail(
+                        d["image"],
+                        d["type"],
+                        d["id"],
+                        "image",
+                        load_image=True
+                    )
+                    self._thumb_map[uid] = {"item": item}
+
+                self.model().appendRow(item)
+
+    def _on_select(self, model_index):
+        """
+        Fires when an item in the completer is selected.
+        This will emit an entity_selected signal for the
+        global search widget
+
+        :param model_index: QModelIndex describing the current item
+        """
+        data = self.get_result(model_index)
+        if data:
+            self.entity_selected.emit(data["type"], data["id"])
+            self.entity_activated.emit(data["type"], data["id"], data["name"])
```

Comment:
The one with 2 args is the one that existed before in the widget. I kept it for backward compatibility (see the docs way up at the top). After a chat with @manneohrstrom I added the second signal with the additional arg to make the returned data easier to display without doing another round trip to SG if it wasn't absolutely necessary.